c/common/src/codingstandards/c/OutOfBounds.qll

```diff
@@ -0,0 +1,772 @@
+    exists(PointerToObjectSourceOrSizeToBufferAccessFunctionConfig config, Expr useOrChild |
+      (
+        useOrChild = use
+        or
+        getArithmeticOffsetValue(use) > 0 and
+        useOrChild = use.getAChild*()
+      ) and
+      config.hasFlow(DataFlow::exprNode(source), DataFlow::exprNode(useOrChild))
+    )
+  }
+
+  predicate bufferUseComputableBufferSize(Expr bufferUse, Expr source, int size) {
+    // flow from a PointerToObjectSource for which we can compute the exact size
+    size = source.(PointerToObjectSource).getFixedSize() and
+    hasFlowFromBufferOrSizeExprToUse(source, bufferUse)
+  }
+
+  predicate bufferUseNonComputableSize(Expr bufferUse, Expr source) {
+    not bufferUseComputableBufferSize(bufferUse, source, _) and
+    hasFlowFromBufferOrSizeExprToUse(source.(DynamicAllocationSource), bufferUse)
+  }
+
+  predicate sizeExprComputableSize(Expr sizeExpr, Expr source, int size) {
+    // computable direct value
+    size = getStatedValue(sizeExpr) and
+    source = sizeExpr
+    or
+    // computable source value that flows to the size expression
+    size = source.(DynamicAllocationSource).getFixedSize() and
+    hasFlowFromBufferOrSizeExprToUse(source.(DynamicAllocationSource).getSizeExprSource(_, _),
+      sizeExpr)
+  }
+
+  int getArithmeticOffsetValue(Expr expr) {
+    result = getStatedValue(expr.(PointerArithmeticExpr).getOperand())
+    or
+    // edge-case: &(array[index]) expressions
+    result = getStatedValue(expr.(AddressOfExpr).getOperand().(PointerArithmeticExpr).getOperand())
+    or
+    // AddExpr
+    result = getStatedValue(expr.(AddExpr).getAnOperand())
+    or
+    // SubExpr
+    result = -getStatedValue(expr.(SubExpr).getAnOperand())
+    or
+    // fall-back
+    not expr instanceof PointerArithmeticExpr and
+    not expr.(AddressOfExpr).getOperand() instanceof PointerArithmeticExpr and
+    result = 0
+  }
+
+  /**
+   * If the size is not computable locally, then it is either:
+   *
+   * 1. A dynamic allocation, from which we can get `getSizeExprSource()', from which
+   *    we can either check specific logic (e.g. string length with offset) or compare GVNs.
+   * 2. An unrelateable size expression, which we might, however, be able to compute the bounds
+   *    of and check against the buffer size, if that is known.
+   *
+   * In case 2, this predicate does not hold.
+   *
+   * NOTE: This predicate does not actually perform the above mentioned heuristics.
+   */
+  predicate sizeExprNonComputableSize(
+    Expr bufferSizeArg, Expr alloc, Expr allocSize, Expr allocSizeBase, int offset
+  ) {
+    bufferSizeArg = any(BufferAccessLibraryFunctionCall call).getAnArgument() and
+    not sizeExprComputableSize(bufferSizeArg, alloc, _) and
+    allocSize = alloc.(DynamicAllocationSource).getSizeExprSource(allocSizeBase, offset) and
+    hasFlowFromBufferOrSizeExprToUse(allocSize, bufferSizeArg)
+  }
+
+  predicate isBufferSizeExprGreaterThanSourceSizeExpr(
```

Comment:
If we can add a short example to the qldoc here of what this would find, that would aid with comprehension.

Reply:
qldoc added (predicate has been renamed to `isSizeArgGreaterThanBufferSize`